File: java/GrepStd.java
// jgrep_std - idiomatic single-threaded Java: NIO Files.readAllBytes + manual
// byte-array search (no regex), buffered stdout. Mirrors go/grep.go semantics.
import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.attribute.BasicFileAttributes;

public class GrepStd {
    static byte[] pat;
    static byte[] lpat;
    static boolean patSet;
    static boolean ci;
    static boolean recursive;
    static boolean multi;
    static boolean matched;
    static OutputStream out;
    static byte[] lowbuf = new byte[0]; // reused ASCII-lowercase scratch

    // ASCII-only, length-preserving lowercase (matches grep -iF).
    static void asciiLower(byte[] dst, byte[] src, int len) {
        for (int i = 0; i < len; i++) {
            byte b = src[i];
            if (b >= 'A' && b <= 'Z') dst[i] = (byte) (b + 32);
            else dst[i] = b;
        }
    }

    // index of needle in hay[from:len], or -1. Empty needle returns from.
    static int indexOf(byte[] hay, int len, int from, byte[] needle) {
        int n = needle.length;
        if (n == 0) return from;
        int end = len - n;
        for (int i = from; i <= end; i++) {
            int j = 0;
            while (j < n && hay[i + j] == needle[j]) j++;
            if (j == n) return i;
        }
        return -1;
    }

    static int lastIndexOfNL(byte[] data, int m) {
        for (int i = m - 1; i >= 0; i--) if (data[i] == '\n') return i;
        return -1;
    }

    static int indexOfNL(byte[] data, int from, int len) {
        for (int i = from; i < len; i++) if (data[i] == '\n') return i;
        return -1;
    }

    static void searchFile(String path) throws IOException {
        byte[] data;
        try {
            data = Files.readAllBytes(Paths.get(path));
        } catch (IOException e) {
            return;
        }
        int len = data.length;
        int peek = Math.min(len, 65536);
        for (int i = 0; i < peek; i++) if (data[i] == 0) return; // binary

        byte[] hay = data;
        byte[] needle = pat;
        if (ci) {
            if (lowbuf.length < len) lowbuf = new byte[len];
            asciiLower(lowbuf, data, len);
            hay = lowbuf;
            needle = lpat;
        }
        byte[] pathBytes = path.getBytes(java.nio.charset.StandardCharsets.ISO_8859_1);
        int pos = 0;
        // pos < len (not <=): an empty needle would otherwise yield a phantom
        // match at pos==len (trailing blank line); grep -F "" prints each line once.
        while (pos < len) {
            int m = indexOf(hay, len, pos, needle);
            if (m < 0) break;
            int ls = lastIndexOfNL(data, m) + 1;
            int j = indexOfNL(data, m, len);
            int le = (j >= 0) ? j : len;
            matched = true;
            if (multi) {
                out.write(pathBytes);
                out.write(':');
            }
            out.write(data, ls, le - ls);
            out.write('\n');
            pos = le + 1;
        }
    }

    static void walk(Path dir) {
        try {
            Files.walkFileTree(dir, new java.nio.file.SimpleFileVisitor<Path>() {
                @Override
                public java.nio.file.FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                    if (attrs.isRegularFile()) searchFile(file.toString());
                    return java.nio.file.FileVisitResult.CONTINUE;
                }
                @Override
                public java.nio.file.FileVisitResult visitFileFailed(Path file, IOException exc) {
                    return java.nio.file.FileVisitResult.CONTINUE;
                }
            });
        } catch (IOException e) { /* ignore */ }
    }

    static void usage() {
        System.err.print("usage: jgrep [-r] [-i] PATTERN PATH...\n");
        System.exit(2);
    }

    public static void main(String[] args) throws IOException {
        java.util.ArrayList<String> paths = new java.util.ArrayList<>();
        boolean noMore = false;
        for (String a : args) {
            if (!noMore && a.length() >= 2 && a.charAt(0) == '-') {
                if (a.equals("--")) { noMore = true; continue; }
                for (int k = 1; k < a.length(); k++) {
                    char c = a.charAt(k);
                    if (c == 'i') ci = true;
                    else if (c == 'r') recursive = true;
                    else usage();
                }
            } else if (!patSet) {
                pat = a.getBytes(java.nio.charset.StandardCharsets.ISO_8859_1);
                patSet = true;
            } else {
                paths.add(a);
            }
        }
        if (!patSet || paths.isEmpty()) usage();
        lpat = new byte[pat.length];
        asciiLower(lpat, pat, pat.length);
        multi = recursive || paths.size() > 1;
        out = new BufferedOutputStream(System.out, 1 << 16);

        for (String p : paths) {
            Path pp = Paths.get(p);
            BasicFileAttributes attrs;
            try {
                attrs = Files.readAttributes(pp, BasicFileAttributes.class);
            } catch (IOException e) {
                continue;
            }
            if (attrs.isDirectory()) {
                if (recursive) walk(pp);
            } else {
                searchFile(p);
            }
        }
        out.flush();
        System.exit(matched ? 0 : 1);
    }
}

File: kotlin/grep_std.kt
// ktgrep_std - idiomatic single-threaded Kotlin: walk the filesystem, read each
// file fully into a ByteArray, NUL-check the 64 KB prefix, scan with ByteArray
// indexOf. ASCII-only case folding (byte-exact, matches grep -iF).
import java.io.BufferedOutputStream
import java.io.File
import kotlin.system.exitProcess

private var pat = ByteArray(0)
private var lpat = ByteArray(0)
private var ci = false
private var recursive = false
private var multi = false
private var matched = false
private lateinit var out: BufferedOutputStream
private var lowbuf = ByteArray(0)

// ASCII-only, length-preserving lowercase.
private fun asciiLower(dst: ByteArray, src: ByteArray, n: Int) {
    for (i in 0 until n) {
        val b = src[i]
        dst[i] = if (b >= 'A'.code.toByte() && b <= 'Z'.code.toByte()) (b + 32).toByte() else b
    }
}

// Index of `needle` in hay[from .. hayLen), byte-exact, like bytes.Index.
private fun indexOf(hay: ByteArray, hayLen: Int, needle: ByteArray, from: Int): Int {
    val n = needle.size
    if (n == 0) return from                 // empty needle: matches at `from` (Go semantics)
    if (from + n > hayLen) return -1
    val first = needle[0]
    var i = from
    val last = hayLen - n
    while (i <= last) {
        if (hay[i] == first) {
            var j = 1
            while (j < n && hay[i + j] == needle[j]) j++
            if (j == n) return i
        }
        i++
    }
    return -1
}

private fun lastNl(data: ByteArray, end: Int): Int {
    var i = end - 1
    while (i >= 0) {
        if (data[i] == '\n'.code.toByte()) return i + 1
        i--
    }
    return 0
}

private fun searchFile(path: String) {
    val data: ByteArray = try {
        File(path).readBytes()
    } catch (e: Exception) {
        return
    }
    val len = data.size
    val peek = if (len > 65536) 65536 else len
    for (i in 0 until peek) {
        if (data[i] == 0.toByte()) return // binary
    }
    val hay: ByteArray
    val needle: ByteArray
    if (ci) {
        if (lowbuf.size < len) lowbuf = ByteArray(len)
        asciiLower(lowbuf, data, len)
        hay = lowbuf
        needle = lpat
    } else {
        hay = data
        needle = pat
    }
    // Guard is `pos < len` (not `<=`): with `<=`, an empty needle would emit one
    // spurious blank match past the file's final newline. grep -F treats a
    // trailing '\n' as terminating the last line, so empty pattern == one match
    // per line exactly. A non-empty needle can never match at pos==len anyway.
    var pos = 0
    while (pos < len) {
        val m = indexOf(hay, len, needle, pos)
        if (m < 0) break
        val ls = lastNl(data, m)
        var le = len
        var j = m
        while (j < len) {
            if (data[j] == '\n'.code.toByte()) { le = j; break }
            j++
        }
        matched = true
        if (multi) {
            out.write(path.toByteArray(Charsets.ISO_8859_1))
            out.write(':'.code)
        }
        out.write(data, ls, le - ls)
        out.write('\n'.code)
        pos = le + 1
    }
}

private fun usage(): Nothing {
    System.err.print("usage: ktgrep [-r] [-i] PATTERN PATH...\n")
    exitProcess(2)
}

private fun collect(p: String, files: MutableList<String>) {
    val f = File(p)
    if (!f.exists()) return
    if (f.isDirectory) {
        if (recursive) {
            // Walk without following symlinks; regular files only.
            val stack = ArrayDeque<File>()
            stack.addLast(f)
            while (stack.isNotEmpty()) {
                val d = stack.removeLast()
                val kids = d.listFiles() ?: continue
                for (k in kids) {
                    // listFiles doesn't follow symlink dirs into them unless we recurse;
                    // skip symlinks explicitly.
                    val isSymlink = try {
                        k.canonicalFile != k.absoluteFile
                    } catch (e: Exception) { false }
                    if (isSymlink) continue
                    if (k.isDirectory) stack.addLast(k)
                    else if (k.isFile) files.add(k.path)
                }
            }
        }
    } else if (f.isFile) {
        files.add(p)
    }
}

fun main(args: Array<String>) {
    val paths = ArrayList<String>()
    var patSet = false
    var noMore = false
    for (a in args) {
        if (!noMore && a.length >= 2 && a[0] == '-') {
            if (a == "--") { noMore = true; continue }
            for (idx in 1 until a.length) {
                when (a[idx]) {
                    'i' -> ci = true
                    'r' -> recursive = true
                    else -> usage()
                }
            }
        } else if (!patSet) {
            pat = a.toByteArray(Charsets.ISO_8859_1)
            patSet = true
        } else {
            paths.add(a)
        }
    }
    if (!patSet || paths.isEmpty()) usage()
    lpat = ByteArray(pat.size)
    asciiLower(lpat, pat, pat.size)
    multi = recursive || paths.size > 1
    out = BufferedOutputStream(System.out, 1 shl 16)

    val files = ArrayList<String>()
    for (p in paths) collect(p, files)
    for (fp in files) searchFile(fp)

    out.flush()
    exitProcess(if (matched) 0 else 1)
}
